Escape special characters in regular expressions for comments

bxt opened the pull request:
Escape special characters in regular expressions for comments.

When figuring out if there is already an `eslint-disable` comment on the line before, the extension uses regular expressions to match comments. But comments include characters that appear naturally in regular expressions, like `*`. So the need to be replaced for construction of the regular expression. Otherwise adding the comment can fail when the previous line happens to include ` eslint-disable-next-line` even if not in a comment, or in languages that are not in the pre-defined list and thus use `/**` as a block comment indicator.

The regular expression to escape regular expressions is inspired by [ones commonly found e.g. on StackOverflow](https://stackoverflow.com/a/6969486/935676).

Fixes #1610 see there for more background and explanation.

bxt commented:
@microsoft-github-policy-service agree

dbaeumer commented on server/src/escapeStringRegexp.ts:
```diff
@@ -0,0 +1,3 @@
+export function escapeStringRegexp(string: string) {
```
Can we move this into eslintServer since it is only used there.

bxt replied:
@dbaeumer Sure, I [moved](https://github.com/microsoft/vscode-eslint/compare/a13fb0851d443ebdaddad759d8085d7ec3720982..ed602f24f0667b502ecd9a3bf209130aeb6d56d0) it close to where it's used now.

MariaSolOs commented on server/src/eslintServer.ts:
```diff
@@ -680,6 +680,10 @@ messageQueue.registerRequest(CodeActionRequest.type, async (params) => {
 		return charIndex;
 	}
 
+	function escapeStringRegexp(string: string) {
```
I think that a comment referencing this PR or the issue would be helpful for other contributors to understand why this escaping is necessary.

dbaeumer commented on server/src/eslintServer.ts:
```diff
@@ -680,6 +680,10 @@ messageQueue.registerRequest(CodeActionRequest.type, async (params) => {
 		return charIndex;
 	}
 
+	function escapeStringRegexp(string: string) {
```
Agree with @MariaSolOs here. I usually call this `value` in these cases.

dbaeumer reviewed (changes requested):
See the comment about renaming `string` to `value`

wesleybl commented:
@bxt could you please make the requested corrections so we can merge. It would be great to have this fix.

bxt commented:
Hey @dbaeumer @MariaSolOs I did the changes as you suggested, renamed the parameter `string` to `value` and added an explanatory comment with a link to the issue. Could you have another look?

vrubezhny commented:
@dbaeumer Any chance to see this PR merged?

bxt commented:
it seems for the "Community PR Approvals" to become green I need one more review, I'm not sure who is eligible, but I see @aeschli is doing reviews here, maybe you could have look as well and review/merge this?

ptrxyz commented:
Yep, a single more review for these few lines and we would be back on track. I find @joaomoreno also doing some reviews, can we get a quick peek from either of you? :)

AndreMaz commented:
@bxt @aeschli @dbaeumer I'm getting 

```bash
[Info  - 1:57:00 PM] ESLint server is starting.
[Info  - 1:57:01 PM] ESLint server running in node v20.9.0
[Info  - 1:57:01 PM] ESLint server is running.
[Info  - 1:48:05 PM] ESLint library loaded from: /home/Dev/web/node_modules/eslint/lib/api.js
[Info  - 1:48:05 PM] Server process exited successfully
[Error - 1:55:18 PM] Request textDocument/codeAction failed.
  Message: Request textDocument/codeAction failed with message: Range#create called with invalid arguments[[object Object], [object Object], undefined, undefined]
  Code: -32603 
  ````

when I create new file in a [t3-turbo](https://github.com/t3-oss/create-t3-turbo) based template.

Running `ESLint: Restart ESLint Server` solves the issue, ie, the I see no error and linting works as expected.


Context:
Plugin
```bash
ESLint: v3.0.10
```

vscode:
```bash
Version: 1.91.0
Commit: ea1445cc7016315d0f5728f8e8b12a45dc0a7286
Date: 2024-07-01T18:52:41.959Z
Electron: 29.4.0
ElectronBuildId: 9728852
Chromium: 122.0.6261.156
Node.js: 20.9.0
V8: 12.2.281.27-electron.0
OS: Linux x64 5.15.0-113-generic
```

dbaeumer commented:
@AndreMaz any steps on how to reproduce this. Otherwise it is hard to tell why it is failing.

soulchild commented:
> when I create new file in a [t3-turbo](https://github.com/t3-oss/create-t3-turbo) based template.
> Running `ESLint: Restart ESLint Server` solves the issue, ie, I see no error and linting works as expected.

Seeing the exact same thing as well. Every time I create a new file in my create-t3-turbo repo, all hell breaks lose and the ESLint server crashes with that error message. A restart via the command palette fixes the issue. Did you find a root cause and solution for this?

AndreMaz commented:
Hey @soulchild 
> Did you find a root cause and solution for this?

No, I got re-assigned to different stuff and didn't had time to investigate. 

Would you mind creating a repro repo to make this thing easier to debug?
